engine-agent/src/main/java/io/camunda/zeebe/process/test/engine/agent/RecordStreamSourceWrapper.java

```diff
@@ -2,22 +2,31 @@
 
 import io.camunda.zeebe.process.test.api.RecordStreamSource;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
+import java.util.stream.StreamSupport;
 
 public class RecordStreamSourceWrapper {
 
   private final List<String> mappedRecords = new ArrayList<>();
   private final RecordStreamSource recordStreamSource;
+  private volatile long lastEventPosition = -1L;
 
   public RecordStreamSourceWrapper(final RecordStreamSource recordStreamSource) {
     this.recordStreamSource = recordStreamSource;
   }
 
   public List<String> getMappedRecords() {
     synchronized (mappedRecords) {
-      mappedRecords.clear();
-      recordStreamSource.records().forEach(record -> mappedRecords.add(record.toJson()));
+      StreamSupport.stream(recordStreamSource.records().spliterator(), false)
```

Comment:
:question: How does this work together with resetting the engine? Wouldn't resetting the engine also reset the record position to 0?

Reply:
Yes it will. Resetting doesn't really reset the engine. It stops the old engine and recreate a new one. This also means there will be a fresh record stream, containing nothing.

Reply:
If that is the case, shouldn't we also reset the RecordStreamWrapper when we reset the engine?

Reply:
A new instance of this class is created upon reseting, so there is no need to do this separately:
```java
@Override
  public void resetEngine(
      final ResetEngineRequest request,
      final StreamObserver<ResetEngineResponse> responseObserver) {
    engine.stop();
    engine = EngineFactory.create(AgentProperties.getGatewayPort());
    recordStreamSource = new RecordStreamSourceWrapper(engine.getRecordStreamSource());
    engine.start();

    final ResetEngineResponse response = ResetEngineResponse.newBuilder().build();
    responseObserver.onNext(response);
    responseObserver.onCompleted();
  }
```